File: php/utility/PrepareAuth.php
<?php
declare(strict_types=1);

// YahooFinance SDK utility: prepare_auth

class YahooFinancePrepareAuth
{
    private const COOKIE_AUTH = 'Session';
    private const HEADER_COOKIE = 'cookie';
    private const OPTION_APIKEY = 'apikey';
    private const NOT_FOUND = '__NOTFOUND__';

    public static function call(YahooFinanceContext $ctx): array
    {
        $spec = $ctx->spec;
        if (!$spec) {
            return [null, $ctx->make_error('auth_no_spec', 'Expected context spec property to be defined.')];
        }

        $headers = &$spec->headers;
        $options = $ctx->client->options_map();

        // Public APIs that need no auth omit the options.auth block entirely.
        if (!isset($options['auth']) || $options['auth'] === null) {
            return [$spec, null];
        }

        $apikey = \Voxgig\Struct\Struct::getprop($options, self::OPTION_APIKEY, self::NOT_FOUND);

        $missing = (is_string($apikey) && ($apikey === self::NOT_FOUND || $apikey === ''))
            || $apikey === null;

        if (!$missing) {
            // A COOKIE IS APPENDED, NEVER ASSIGNED: the header may already
            // carry the caller's own cookies, and one `Cookie:` header
            // holds all of them, separated by '; '.
            $apikey_val = is_string($apikey) ? $apikey : '';
            $existing = $headers[self::HEADER_COOKIE] ?? '';
            $pair = self::COOKIE_AUTH . '=' . $apikey_val;
            $headers[self::HEADER_COOKIE] = $existing === ''
                ? $pair : "{$existing}; {$pair}";
        }

        return [$spec, null];
    }
}

File: lua/utility/prepare_auth.lua
-- YahooFinance SDK utility: prepare_auth

local vs = require("utility.struct.struct")

local COOKIE_AUTH = "Session"
local HEADER_COOKIE = "cookie"
local OPTION_APIKEY = "apikey"
local NOT_FOUND = "__NOTFOUND__"

local function cookie_set(headers, value)
  local kept = {}
  local existing = headers[HEADER_COOKIE]

  if type(existing) == "string" then
    for pair in string.gmatch(existing, "[^;]+") do
      local one = string.match(pair, "^%s*(.-)%s*$")
      if one ~= "" and string.match(one, "^[^=]*") ~= COOKIE_AUTH then
        kept[#kept + 1] = one
      end
    end
  end

  if value ~= nil then
    kept[#kept + 1] = COOKIE_AUTH .. "=" .. value
  end

  if #kept == 0 then
    headers[HEADER_COOKIE] = nil
  else
    headers[HEADER_COOKIE] = table.concat(kept, "; ")
  end
end


local function prepare_auth_util(ctx)
  local spec = ctx.spec
  if spec == nil then
    return nil, ctx:make_error("auth_no_spec",
      "Expected context spec property to be defined.")
  end

  local headers = spec.headers
  local options = ctx.client:options_map()

  -- Public APIs that need no auth omit the options.auth block entirely.
  if options.auth == nil then
    cookie_set(headers, nil)
    return spec, nil
  end

  local apikey = vs.getprop(options, OPTION_APIKEY, NOT_FOUND)

  if apikey == nil
    or (type(apikey) == "string" and (apikey == NOT_FOUND or apikey == ""))
  then
    cookie_set(headers, nil)
  else
    local apikey_val = ""
    if type(apikey) == "string" then
      apikey_val = apikey
    end
    -- The prefix is a header-value convention and has no meaning in a
    -- cookie pair, so it is dropped the way a query placement drops it.
    cookie_set(headers, apikey_val)
  end

  return spec, nil
end

return prepare_auth_util
